Task4.1/MatrixTracker.cs

```diff
@@ -0,0 +1,37 @@
+﻿using static System.Console;
+
+namespace Task41
+{
+    public class MatrixTracker<T>
+    {
+        private readonly DiagonalMatrix<T> matrix;
+        private readonly Stack<(int Row, int Column, T OldValue)> changeHistory = new Stack<(int, int, T)>();
+
+        public MatrixTracker(DiagonalMatrix<T> matrix)
+        {
+            this.matrix = matrix ?? throw new ArgumentNullException(nameof(matrix));
+            this.matrix.ElementChanged += OnElementChanged;
+        }
+
+        private void OnElementChanged(object sender, ElementChangedEventArgs<T> e)
+        {
+            // Track changes in history
+            changeHistory.Push((e.Row, e.Column, e.OldValue));
+            WriteLine($"Element at ({e.Row}, {e.Column}) changed from {e.OldValue} to {e.NewValue}");
+        }
+
+        // Undo the last change
+        public void Undo()
+        {
+            if (changeHistory.Count > 0)
+            {
```

Comment:
Unsubcribe before this operation and subsribe again after it, not to have a loop of changes